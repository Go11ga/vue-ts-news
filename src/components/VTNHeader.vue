<template>
    <div class="header">

        <div class="header__item-about">
            <h1 class="header__title">
                Список новостей
            </h1>

            <button class="header__refresh-btn" @click="updateNews">
                <img class="header__refresh-img" src="@/assets/images/refresh.png" alt="refresh">
            </button>
        </div>

        <div class="header__item-search">   
            <label class="header__input-label">
                <input 
                    class="header__input" 
                    v-model.trim="searchValue" 
                    @keyup.enter="searchNews"
                >
                <button class="header__search-btn" @click="searchNews">
                    <img class="header__preloader-img" src="@/assets/images/search_preloader.gif" v-if="news.newsProcess">
                    <img class="header__search-img" src="@/assets/images/search.png" v-else>
                </button>
            </label>
        </div>

    </div>
</template>

<script lang="ts">

import { defineComponent } from 'vue'
import { mapState, mapMutations, mapActions } from 'vuex'

export default defineComponent({
    name: 'VTNHeader',

    data() {
        return {
            searchValue: ''
        }
    },

    computed: {
        ...mapState([
            'news'
        ])
    },

    methods: {
        ...mapMutations([
            'CLEAR_NEWS',
            'SET_SEARCH_VALUE'
        ]),

        ...mapActions([
            'fetchNews'
        ]),

        updateNews() {
            this.CLEAR_NEWS()
            this.fetchNews()
        },

        searchNews() {
            this.SET_SEARCH_VALUE(this.searchValue)
            this.fetchNews()
        }
    }
})
</script>
