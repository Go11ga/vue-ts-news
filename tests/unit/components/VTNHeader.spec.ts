import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'
import VTNHeader from '@/components/VTNHeader.vue'

const makeWrapper = (newsProcess: boolean = false) => {
    const state = () => {
        return {
            news: {
                newsProcess
            }
        }
    }

    const store = createStore({
        state
    })

    const wrapper = shallowMount(VTNHeader, {
        global: {
            plugins: [store]
        }
    })

    return wrapper
}

describe('VTNHeader', () => {
    let wrapper: any

    beforeEach(() => {
        wrapper = makeWrapper()

        wrapper.vm.CLEAR_NEWS = jest.fn()
        wrapper.vm.SET_SEARCH_VALUE = jest.fn()
        wrapper.vm.fetchNews = jest.fn()
    })

    afterEach(() => {
        wrapper.unmount()
    })

    it('VTNHeader dummy name test', () => {
        expect(wrapper.vm.$options.name).toMatch('VTNHeader')
    })

    it('заголовок h1', () => {
        expect(wrapper.find('h1').text()).toMatch('Список новостей')
    })

    it('иконки в поле поиска', async () => {
        expect(wrapper.html()).toContain('header__search-img')
        expect(wrapper.html()).not.toContain('header__preloader-img')

        wrapper = makeWrapper(true)

        await wrapper.vm.$nextTick()

        expect(wrapper.html()).not.toContain('header__search-img')
        expect(wrapper.html()).toContain('header__preloader-img')
    })

    it('snapshot', () => {
        expect(wrapper.element).toMatchSnapshot()
    })

    it('getters', () => {
        expect(wrapper.vm.news).toEqual({ newsProcess: false })

        expect(wrapper.vm.searchValue).toBe('')
        
    })

    it('клик по кнопке refresh', () => {
        let spy = jest.spyOn(wrapper.vm, 'updateNews')

        wrapper.find('button.header__refresh-btn').trigger('click')
        
        expect(spy).toHaveBeenCalled()
        expect(spy).toHaveBeenCalledTimes(1)
    })

    it('метод updateNews', () => {
        wrapper.vm.updateNews()

        expect(wrapper.vm.CLEAR_NEWS).toHaveBeenCalled()
        expect(wrapper.vm.fetchNews).toHaveBeenCalled()
    })

    it('v-model .header__input', () => {
        wrapper.find('.header__input').setValue('test')

        expect(wrapper.vm.searchValue).toBe('test')
    })

    it('enter в input', () => {
        wrapper.vm.searchNews = jest.fn()

        wrapper.find('.header__input').trigger('keyup.enter')

        expect(wrapper.vm.searchNews).toHaveBeenCalled()
    })

    it('клик по кнопке search', () => {
        wrapper.vm.searchNews = jest.fn()

        wrapper.find('.header__search-btn').trigger('click')
        
        expect(wrapper.vm.searchNews).toHaveBeenCalled()
    })

    it('метод searchNews', async () => {
        wrapper.setData({ searchValue: 'test' })

        wrapper.vm.searchNews()

        await wrapper.vm.$nextTick()

        expect(wrapper.vm.SET_SEARCH_VALUE).toHaveBeenCalledWith('test')
        expect(wrapper.vm.fetchNews).toHaveBeenCalled()
    })
})
