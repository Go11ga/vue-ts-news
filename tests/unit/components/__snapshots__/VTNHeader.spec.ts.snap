// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTNHeader snapshot 1`] = `
<div
  class="header"
>
  <div
    class="header__item-about"
  >
    <h1
      class="header__title"
    >
       Список новостей 
    </h1>
    <button
      class="header__refresh-btn"
    >
      <img
        alt="refresh"
        class="header__refresh-img"
      />
    </button>
  </div>
  <div
    class="header__item-search"
  >
    <label
      class="header__input-label"
    >
      <input
        class="header__input"
      />
      <button
        class="header__search-btn"
      >
        <img
          class="header__search-img"
        />
      </button>
    </label>
  </div>
</div>
`;
